// @progress/kendo-vue-data-tools/dist/npm/filter/filters/EnumFilter.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];

      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }

    return t;
  };

  return __assign.apply(this, arguments);
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnumFilterVue2 = exports.EnumFilter = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';

var kendo_vue_dropdowns_1 = require("@progress/kendo-vue-dropdowns");

var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");

var messages_1 = require("./../../messages");
/**
 * @hidden
 */


var EnumFilterVue2 = {
  name: 'KendoEnumFilter',
  props: {
    filter: {
      type: Object,
      required: true
    },
    dataItems: Array,
    defaultItem: Object
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var locService = kendo_vue_intl_1.provideLocalizationService(this);
    var _a = this.$props,
        filter = _a.filter,
        _b = _a.dataItems,
        dataItems = _b === void 0 ? [] : _b,
        defaultItem = _a.defaultItem;
    var locData = dataItems.map(function (item) {
      return __assign(__assign({}, item), {
        text: locService.toLanguageString(item.text, messages_1.messages[item.text] || item.text)
      });
    });
    return (// @ts-ignore
      h(kendo_vue_dropdowns_1.DropDownList, {
        value: locData.find(function (i) {
          return i.value === filter.value;
        }) || null,
        attrs: this.v3 ? undefined : {
          value: locData.find(function (i) {
            return i.value === filter.value;
          }) || null,
          defaultItem: defaultItem,
          dataItems: locData,
          textField: "text"
        },
        onChange: this.onChange,
        on: this.v3 ? undefined : {
          "change": this.onChange
        },
        defaultItem: defaultItem,
        dataItems: locData,
        textField: "text"
      })
    );
  },
  methods: {
    onChange: function onChange(event) {
      this.$emit('filterchange', {
        nextFilter: __assign(__assign({}, this.$props.filter), {
          value: event.value.value
        })
      });
    }
  }
};
exports.EnumFilterVue2 = EnumFilterVue2;
/**
 * @hidden
 */

var EnumFilter = EnumFilterVue2;
exports.EnumFilter = EnumFilter;
