// @progress/kendo-vue-labels/dist/npm/FloatingLabel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FloatingLabelVue2 = exports.FloatingLabel = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
var inject = allVue.inject;

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");

var messages_1 = require("./messages");

var package_metadata_1 = require("./package-metadata");
/**
 * @hidden
 */


var FloatingLabelVue2 = {
  name: 'KendoFloatingLabel',
  props: {
    label: String,
    editorId: String,
    editorValue: [String, Boolean, Number],
    editorPlaceholder: String,
    editorValid: {
      type: Boolean,
      default: undefined
    },
    editorDisabled: Boolean,
    id: String,
    optional: Boolean
  },
  data: function data() {
    return {
      focused: false
    };
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  created: function created() {
    kendo_vue_common_1.validatePackage(package_metadata_1.packageMetadata);
  },
  computed: {
    spanClassNames: function spanClassNames() {
      return {
        'k-floating-label-container': true,
        'k-focus': this.focused,
        'k-empty': !this.$props.editorPlaceholder && !this.$props.editorValue && this.$props.editorValue !== 0,
        'k-text-disabled': this.$props.editorDisabled,
        'k-rtl': this.$props.dir === 'rtl'
      };
    },
    labelClassNames: function labelClassNames() {
      return {
        'k-label': true,
        'k-text-error': this.$props.editorValid === false,
        'k-text-disabled': this.$props.editorDisabled
      };
    }
  },
  methods: {
    handleFocus: function handleFocus() {
      this.focused = true;
    },
    handleBlur: function handleBlur() {
      this.focused = false;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var defaultSlots = kendo_vue_common_1.getDefaultSlots(this);
    var _a = this.$props,
        label = _a.label,
        editorId = _a.editorId,
        id = _a.id,
        optional = _a.optional;
    var ls = kendo_vue_intl_1.provideLocalizationService(this);
    var localizedOptional = optional ? ls.toLanguageString(messages_1.labelsOptional, messages_1.messages[messages_1.labelsOptional]) : '';
    var optionalElement = localizedOptional && h("span", {
      "class": 'k-label-optional'
    }, [localizedOptional]);
    return h("span", {
      "class": this.spanClassNames,
      onFocusin: this.handleFocus,
      on: this.v3 ? undefined : {
        "focusin": this.handleFocus,
        "focusout": this.handleBlur
      },
      onFocusout: this.handleBlur,
      dir: this.$props.dir,
      attrs: this.v3 ? undefined : {
        dir: this.$props.dir
      }
    }, [defaultSlots, label ? editorId ? h("label", {
      id: id,
      attrs: this.v3 ? undefined : {
        id: id,
        "for": editorId
      },
      "for": editorId,
      "class": this.labelClassNames
    }, [label, optionalElement]) : h("span", {
      id: id,
      attrs: this.v3 ? undefined : {
        id: id
      },
      "class": this.labelClassNames
    }, [label, optionalElement]) : null]);
  }
};
exports.FloatingLabelVue2 = FloatingLabelVue2;
/**
 * @hidden
 */

var FloatingLabel = FloatingLabelVue2;
exports.FloatingLabel = FloatingLabel;
