// @progress/kendo-vue-grid/dist/npm/cells/GridHierarchyCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridHierarchyCellVue2 = exports.GridHierarchyCell = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';

var utils_1 = require("./../utils");

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var kendo_vue_data_tools_1 = require("@progress/kendo-vue-data-tools");
/**
 * @hidden
 */


var GridHierarchyCellVue2 = {
  props: {
    id: String,
    field: String,
    dataItem: Object,
    format: String,
    type: String,
    className: String,
    colSpan: Number,
    columnIndex: Number,
    columnsCount: Number,
    rowType: String,
    level: Number,
    expanded: Boolean,
    editor: String,
    isSelected: Boolean,
    dataIndex: Number,
    ariaColumnIndex: Number,
    render: [String, Function, Object]
  },
  // @ts-ignore
  emits: {
    change: null,
    cellkeydown: null
  },
  inject: {
    kendoIntlService: {
      default: null
    },
    getKeyboardNavigationAttributes: {
      default: kendo_vue_common_1.noop
    }
  },
  methods: {
    triggerKeydown: function triggerKeydown(event, expanded) {
      this.$emit('cellkeydown', {
        event: event,
        dataItem: this.$props.dataItem,
        field: this.$props.field,
        expanded: expanded
      });

      if (event.defaultPrevented) {
        return;
      }

      if (event.keyCode === kendo_vue_common_1.Keys.enter) {
        event.preventDefault();
        this.$emit('change', {
          dataItem: this.$props.dataItem,
          dataIndex: this.$props.dataIndex,
          event: event,
          field: this.$props.field,
          value: !expanded
        });
      }
    },
    clickHandler: function clickHandler(e, dataItem, expanded) {
      e.preventDefault();
      this.$emit('change', {
        dataItem: dataItem,
        event: e,
        field: undefined,
        value: !expanded
      });
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;

    var h = gh || createElement;
    var defaultRendering = null;
    var navAttrs = this.getKeyboardNavigationAttributes(this.$props.id);

    if (this.$props.rowType === 'groupFooter') {
      defaultRendering = h("td", {
        "class": "k-hierarchy-cell"
      });
    } else if (this.$props.rowType !== 'groupHeader') {
      var expanded_1 = utils_1.getNestedValue(this.$props.field, this.$props.dataItem);
      var className = expanded_1 ? 'k-icon k-i-minus' : 'k-icon k-i-plus';
      defaultRendering = h("td", {
        onKeydown: function onKeydown(ev) {
          _this.triggerKeydown(ev, expanded_1);
        },
        on: this.v3 ? undefined : {
          "keydown": function onKeydown(ev) {
            _this.triggerKeydown(ev, expanded_1);
          }
        },
        "class": "k-hierarchy-cell",
        "aria-expanded": expanded_1 ? 'true' : 'false',
        attrs: this.v3 ? undefined : {
          "aria-expanded": expanded_1 ? 'true' : 'false',
          role: 'gridcell',
          "aria-colindex": this.$props.ariaColumnIndex,
          tabindex: navAttrs.tabIndex,
          "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
          "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
        },
        role: 'gridcell',
        "aria-colindex": this.$props.ariaColumnIndex,
        tabindex: navAttrs.tabIndex,
        "data-keyboardnavlevel": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_LEVEL],
        "data-keyboardnavid": navAttrs[kendo_vue_data_tools_1.KEYBOARD_NAV_DATA_ID]
      }, [h("a", {
        onClick: function onClick(e) {
          _this.clickHandler(e, _this.$props.dataItem, expanded_1);
        },
        on: this.v3 ? undefined : {
          "click": function onClick(e) {
            _this.clickHandler(e, _this.$props.dataItem, expanded_1);
          }
        },
        "class": className,
        href: "#",
        attrs: this.v3 ? undefined : {
          href: "#",
          tabindex: -1
        },
        tabindex: -1
      })]);
    }

    return kendo_vue_common_1.getTemplate.call(this, {
      h: h,
      template: this.$props.render,
      defaultRendering: defaultRendering,
      additionalProps: this.$props,
      additionalListeners: {
        keydown: this.triggerKeydown,
        click: this.clickHandler
      }
    });
  }
};
exports.GridHierarchyCellVue2 = GridHierarchyCellVue2;
/**
 * @hidden
 */

var GridHierarchyCell = GridHierarchyCellVue2;
exports.GridHierarchyCell = GridHierarchyCell;
