// @progress/kendo-vue-grid/dist/es/footer/Footer.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
var ref = allVue.ref;
import { BrowserSupportService } from '../utils/browser-support.service';
import { isRtl } from '../utils';
import { getRef, setRef } from '@progress/kendo-vue-common';
/**
 * @hidden
 */

var FooterVue2 = {
  name: 'Footer',
  props: {
    staticHeaders: Boolean,
    row: Object,
    columnResize: Object,
    cols: Array
  },
  data: function data() {
    return {
      scrollbarWidth: 0,
      rtl: false,
      tableWidth: null
    };
  },
  computed: {
    divStyle: {
      get: function get() {
        var padding = this.$data.scrollbarWidth + 'px';
        var right = this.rtl ? 0 : padding;
        var left = this.rtl ? padding : 0;
        return {
          padding: "0 " + right + " 0 " + left
        };
      }
    },
    tableStyle: {
      get: function get() {
        return this.tableWidth ? {
          width: this.tableWidth
        } : null;
      }
    }
  },
  mounted: function mounted() {
    this.$data.scrollbarWidth = new BrowserSupportService().scrollbarWidth;
    this.rtl = isRtl(this.$el);
    this.$props.columnResize.colGroupFooter = getRef(this, 'colGroupHeader');
    this._footerWrap = getRef(this, 'footerWrap');
  },
  methods: {
    setScrollLeft: function setScrollLeft(scrollLeft) {
      if (this._footerWrap) {
        this._footerWrap.scrollLeft = scrollLeft;
      }
    },
    setWidth: function setWidth(width) {
      this.$data.tableWidth = width + 'px';
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var footerWrapRef = ref(null);
    var colGroupHeaderRef = ref(null);
    return {
      v3: v3,
      colGroupHeaderRef: colGroupHeaderRef,
      footerWrapRef: footerWrapRef
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;

    if (!this.$props.staticHeaders) {
      return h("tfoot", {
        "class": "k-grid-footer"
      }, [this.$props.row]);
    }

    return h("div", {
      "class": "k-grid-footer",
      style: this.divStyle
    }, [h("div", {
      "class": "k-grid-footer-wrap",
      ref: setRef(this, 'footerWrap')
    }, [h("table", {
      style: this.tableStyle
    }, [h("colgroup", {
      ref: setRef(this, 'colGroupHeader')
    }, [this.$props.cols]), h("tfoot", [this.$props.row])])])]);
  }
};
/**
 * @hidden
 */

var Footer = FooterVue2;
export { Footer, FooterVue2 };
