// @progress/kendo-vue-data-tools/dist/es/pager/PagerPageSizes.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
var inject = allVue.inject;
import { messages, pagerItemPerPage } from './../messages';
import { provideLocalizationService } from '@progress/kendo-vue-intl';
import { DropDownList } from '@progress/kendo-vue-dropdowns';
/**
 * @hidden
 */

var PagerPageSizesVue2 = {
  name: 'KendoPagerPageSizes',
  props: {
    pageSize: Number,
    pageSizes: Array,
    messagesMap: Function
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  methods: {
    pageSizeChange: function pageSizeChange(e) {
      this.$emit('pagechange', {
        skip: 0,
        take: parseInt(e.target.value, 10)
      }, e);
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var _this = this;

    var h = gh || createElement;
    var sizes = this.$props.pageSizes.slice();

    if (sizes.filter(function (s) {
      return s === _this.$props.pageSize;
    }).length === 0) {
      sizes.unshift(this.$props.pageSize);
    }

    var itemPerPageMessage = this.$props.messagesMap ? this.$props.messagesMap(pagerItemPerPage) : {
      messageKey: pagerItemPerPage,
      defaultMessage: messages[pagerItemPerPage]
    };
    return h("span", {
      "class": "k-pager-sizes k-label"
    }, [// @ts-ignore function children
    h(DropDownList, {
      value: this.$props.pageSize,
      attrs: this.v3 ? undefined : {
        value: this.$props.pageSize,
        dataItems: sizes
      },
      dataItems: sizes,
      onChange: this.pageSizeChange,
      on: this.v3 ? undefined : {
        "change": this.pageSizeChange
      }
    }), provideLocalizationService(this).toLanguageString(itemPerPageMessage.messageKey, itemPerPageMessage.defaultMessage)]);
  }
};
/**
 * @hidden
 */

var PagerPageSizes = PagerPageSizesVue2;
export { PagerPageSizes, PagerPageSizesVue2 };
