// @progress/kendo-vue-grid/dist/es/drag/DragClue.js
// @ts-ignore
import * as Vue from 'vue';
var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
/**
 * @hidden
 */

var DragClueVue2 = {
  data: function data() {
    return {
      visible: false,
      top: 0,
      left: 0,
      innerText: '',
      status: 'k-i-cancel'
    };
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    return this.visible && h("div", {
      "class": "k-header k-drag-clue",
      style: {
        display: 'block',
        position: 'absolute',
        zIndex: 20000,
        padding: '8px 12px',
        top: this.top + 'px',
        left: this.left + 'px'
      }
    }, [h("span", {
      "class": 'k-icon k-drag-status ' + this.status + ' k-icon-with-modifier'
    }, [h("span", {
      "class": "k-icon k-icon-modifier"
    })]), this.innerText]);
  }
};
/**
 * @hidden
 */

var DragClue = DragClueVue2;
export { DragClue, DragClueVue2 };
