// @progress/kendo-vue-inputs/dist/npm/colors/ColorContrastLabels.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColorContrastLabelsVue2 = exports.ColorContrastLabels = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';

var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");

var messages_1 = require("../messages");

var color_parser_1 = require("./utils/color-parser");
/**
 * @hidden
 */


var ColorContrastLabelsVue2 = {
  name: 'KendoColorContrastLabels',
  props: {
    bgColor: Object,
    rgba: Object
  },
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var ls = kendo_vue_intl_1.provideLocalizationService(this);
    var contrastRatioMessage = ls.toLanguageString(messages_1.colorGradientContrastRatio, messages_1.messages[messages_1.colorGradientContrastRatio]);
    var AALevelMessage = ls.toLanguageString(messages_1.colorGradientAALevel, messages_1.messages[messages_1.colorGradientAALevel]);
    var AAALevelMessage = ls.toLanguageString(messages_1.colorGradientAAALevel, messages_1.messages[messages_1.colorGradientAAALevel]);
    var passMessage = ls.toLanguageString(messages_1.colorGradientPass, messages_1.messages[messages_1.colorGradientPass]);
    var failMessage = ls.toLanguageString(messages_1.colorGradientFail, messages_1.messages[messages_1.colorGradientFail]);
    var contrast = color_parser_1.getContrastFromTwoRGBAs(this.$props.rgba, this.$props.bgColor);
    var requiredAAContrast = 4.5.toFixed(1);
    var requiredAAAContrast = 7.0.toFixed(1);
    var contrastText = contrastRatioMessage + ": " + contrast.toFixed(2);
    var aaText = AALevelMessage + ": " + requiredAAContrast;
    var aaaText = AAALevelMessage + ": " + requiredAAAContrast;
    var success = h("span", {
      "class": "k-contrast-validation k-text-success"
    }, [passMessage, " ", h("span", {
      "class": "k-icon k-i-check"
    })]);
    var error = h("span", {
      "class": "k-contrast-validation k-text-error"
    }, [failMessage, " ", h("span", {
      "class": "k-icon k-i-close"
    })]);
    return h("div", {
      "class": "k-vbox k-colorgradient-color-contrast"
    }, [h("div", {
      "class": "k-contrast-ratio"
    }, [h("span", {
      "class": "k-contrast-ratio-text"
    }, [contrastText]), contrast >= 4.5 ? h("span", {
      "class": "k-contrast-validation k-text-success"
    }, [h("span", {
      "class": "k-icon k-i-check"
    }), contrast >= 7 && h("span", {
      "class": "k-icon k-i-check"
    })]) : h("span", {
      "class": "k-contrast-validation k-text-error"
    }, [h("span", {
      "class": "k-icon k-i-close"
    })])]), h("div", [h("span", [aaText]), contrast >= 4.5 ? success : error]), h("div", [h("span", [aaaText]), contrast >= 7 ? success : error])]);
  }
};
exports.ColorContrastLabelsVue2 = ColorContrastLabelsVue2;
/**
 * @hidden
 */

var ColorContrastLabels = ColorContrastLabelsVue2;
exports.ColorContrastLabels = ColorContrastLabels;
