// @progress/kendo-vue-inputs/dist/npm/colors/HexInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HexInputVue2 = exports.HexInput = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';

var color_parser_1 = require("./utils/color-parser");

var misc_1 = require("./utils/misc");

var Input_1 = require("../input/Input");
/**
 * @hidden
 */


var HexInputVue2 = {
  name: 'KendoHexInput',
  // @ts-ignore
  emits: {
    'hexchange': null,
    'blur': null,
    'focus': null
  },
  props: {
    hex: String,
    disabled: Boolean
  },
  data: function data() {
    return {
      currentHex: this.$props.hex,
      originalHex: this.$props.hex
    };
  },
  computed: {
    computedHex: function computedHex() {
      return this.hex !== undefined ? this.hex : this.currentHex;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    return h(Input_1.Input, {
      value: this.computedHex,
      attrs: this.v3 ? undefined : {
        value: this.computedHex,
        disabled: this.$props.disabled
      },
      onChange: this.onChange,
      on: this.v3 ? undefined : {
        "change": this.onChange,
        "focus": this.onFocus,
        "blur": this.onBlur
      },
      onFocus: this.onFocus,
      onBlur: this.onBlur,
      disabled: this.$props.disabled
    });
  },
  methods: {
    onChange: function onChange(event) {
      var hex = event.target.value;
      var value = color_parser_1.parseColor(hex, 'rgba');
      this.currentHex = hex;

      if (misc_1.isPresent(value)) {
        this.$emit('hexchange', hex, value, event);
      }
    },
    onBlur: function onBlur(event) {
      if (!misc_1.isPresent(color_parser_1.parseColor(this.hex, 'rgba'))) {
        this.currentHex = this.originalHex;
      }

      this.$emit('blur', event);
    },
    onFocus: function onFocus(event) {
      this.$emit('focus', event);
    }
  }
};
exports.HexInputVue2 = HexInputVue2;
/**
 * @hidden
 */

var HexInput = HexInputVue2;
exports.HexInput = HexInput;
