// @progress/kendo-vue-data-tools/dist/npm/navigation/constants.js
"use strict";

var __spreadArrays = undefined && undefined.__spreadArrays || function () {
  for (var s = 0, i = 0, il = arguments.length; i < il; i++) {
    s += arguments[i].length;
  }

  for (var r = Array(s), k = 0, i = 0; i < il; i++) {
    for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++) {
      r[k] = a[j];
    }
  }

  return r;
};

var _a, _b, _c;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tableKeyboardNavigationBodyAttributes = exports.tableKeyboardNavigationHeaderAttributes = exports.tableKeyboardNavigationScopeAttributes = exports.TABBABLE_ELEMENTS = exports.FOCUSABLE_ELEMENTS = exports.KEYBOARD_NAV_FILTER_COL_SUFFIX = exports.KEYBOARD_NAV_DATA_ZONE = exports.KEYBOARD_NAV_DATA_ID = exports.KEYBOARD_NAV_DATA_BODY = exports.KEYBOARD_NAV_DATA_HEADER = exports.KEYBOARD_NAV_DATA_SCOPE = exports.KEYBOARD_NAV_DATA_LEVEL = void 0;
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_LEVEL = 'data-keyboardnavlevel';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_SCOPE = 'data-keyboardnavscope';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_HEADER = 'data-keyboardnavheader';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_BODY = 'data-keyboardnavbody';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_ID = 'data-keyboardnavid';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_DATA_ZONE = 'data-keyboardnavzone';
/**
 * @hidden
 */

exports.KEYBOARD_NAV_FILTER_COL_SUFFIX = '_filter';
var FOCUSABLE_ELEMENTS_BASE = ['input:not([disabled]):not([type=hidden])', 'select:not([disabled])', 'textarea:not([disabled])', 'button:not([disabled])', 'a[href]', 'area[href]', 'summary', 'iframe', 'object', 'embed', 'audio[controls]', 'video[controls]', '[contenteditable]'];
/**
 * @hidden
 */

exports.FOCUSABLE_ELEMENTS = __spreadArrays(FOCUSABLE_ELEMENTS_BASE, ['[tabindex]']);
/**
 * @hidden
 */

exports.TABBABLE_ELEMENTS = __spreadArrays(FOCUSABLE_ELEMENTS_BASE, ['[tabindex]']).map(function (selector) {
  return selector + ':not([tabindex="-1"])';
});
/**
 * @hidden
 */

exports.tableKeyboardNavigationScopeAttributes = (_a = {}, _a[exports.KEYBOARD_NAV_DATA_SCOPE] = true, _a);
/**
 * @hidden
 */

exports.tableKeyboardNavigationHeaderAttributes = (_b = {}, _b[exports.KEYBOARD_NAV_DATA_HEADER] = true, _b);
/**
 * @hidden
 */

exports.tableKeyboardNavigationBodyAttributes = (_c = {}, _c[exports.KEYBOARD_NAV_DATA_BODY] = true, _c);
