// @progress/kendo-vue-grid/dist/npm/footer/Footer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FooterVue2 = exports.Footer = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
var ref = allVue.ref;

var browser_support_service_1 = require("../utils/browser-support.service");

var utils_1 = require("../utils");

var kendo_vue_common_1 = require("@progress/kendo-vue-common");
/**
 * @hidden
 */


var FooterVue2 = {
  name: 'Footer',
  props: {
    staticHeaders: Boolean,
    row: Object,
    columnResize: Object,
    cols: Array
  },
  data: function data() {
    return {
      scrollbarWidth: 0,
      rtl: false,
      tableWidth: null
    };
  },
  computed: {
    divStyle: {
      get: function get() {
        var padding = this.$data.scrollbarWidth + 'px';
        var right = this.rtl ? 0 : padding;
        var left = this.rtl ? padding : 0;
        return {
          padding: "0 " + right + " 0 " + left
        };
      }
    },
    tableStyle: {
      get: function get() {
        return this.tableWidth ? {
          width: this.tableWidth
        } : null;
      }
    }
  },
  mounted: function mounted() {
    this.$data.scrollbarWidth = new browser_support_service_1.BrowserSupportService().scrollbarWidth;
    this.rtl = utils_1.isRtl(this.$el);
    this.$props.columnResize.colGroupFooter = kendo_vue_common_1.getRef(this, 'colGroupHeader');
    this._footerWrap = kendo_vue_common_1.getRef(this, 'footerWrap');
  },
  methods: {
    setScrollLeft: function setScrollLeft(scrollLeft) {
      if (this._footerWrap) {
        this._footerWrap.scrollLeft = scrollLeft;
      }
    },
    setWidth: function setWidth(width) {
      this.$data.tableWidth = width + 'px';
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    var footerWrapRef = ref(null);
    var colGroupHeaderRef = ref(null);
    return {
      v3: v3,
      colGroupHeaderRef: colGroupHeaderRef,
      footerWrapRef: footerWrapRef
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;

    if (!this.$props.staticHeaders) {
      return h("tfoot", {
        "class": "k-grid-footer"
      }, [this.$props.row]);
    }

    return h("div", {
      "class": "k-grid-footer",
      style: this.divStyle
    }, [h("div", {
      "class": "k-grid-footer-wrap",
      ref: kendo_vue_common_1.setRef(this, 'footerWrap')
    }, [h("table", {
      style: this.tableStyle
    }, [h("colgroup", {
      ref: kendo_vue_common_1.setRef(this, 'colGroupHeader')
    }, [this.$props.cols]), h("tfoot", [this.$props.row])])])]);
  }
};
exports.FooterVue2 = FooterVue2;
/**
 * @hidden
 */

var Footer = FooterVue2;
exports.Footer = Footer;
