// @progress/kendo-vue-dropdowns/dist/npm/MultiSelect/TagList.js
"use strict";

var __assign = undefined && undefined.__assign || function () {
  __assign = Object.assign || function (t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
      s = arguments[i];

      for (var p in s) {
        if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
      }
    }

    return t;
  };

  return __assign.apply(this, arguments);
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TagListVue2 = exports.TagList = void 0; // @ts-ignore

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var isV3 = allVue.version[0] === '3';
var ref = allVue.ref;

var preventDefault = function preventDefault(event) {
  return event.preventDefault();
};

var stopPropagation = function stopPropagation(event) {
  return event.stopPropagation();
};
/**
 * @hidden
 */


var TagListVue2 = {
  name: 'TagList',
  props: {
    dataItems: Array,
    guid: String,
    focused: Object,
    tagRender: [String, Function, Object],
    tagsRounded: {
      type: String,
      default: 'medium',
      validator: function validator(value) {
        return ['small', 'medium', 'large', 'full'].includes(value);
      }
    },
    fillMode: {
      type: String,
      default: 'solid',
      validator: function validator(value) {
        return ['solid', 'flat', 'outline'].includes(value);
      }
    },
    size: {
      type: String,
      default: 'medium',
      validator: function validator(value) {
        return ['small', 'medium', 'large'].includes(value);
      }
    }
  },
  computed: {
    wrapperClass: function wrapperClass() {
      var _a;

      var _b = this.$props,
          size = _b.size,
          disabled = _b.disabled;
      return _a = {
        'k-chip-list': true
      }, _a["k-chip-list-" + (kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size)] = size, _a['k-selection-single'] = true, _a['k-disabled'] = disabled, _a;
    },
    chipClass: function chipClass() {
      var _a;

      var _b = this.$props,
          size = _b.size,
          disabled = _b.disabled,
          tagsRounded = _b.tagsRounded,
          fillMode = _b.fillMode;
      return _a = {
        'k-chip': true,
        'k-disabled': disabled
      }, _a["k-chip-" + (kendo_vue_common_1.kendoThemeMaps.sizeMap[size] || size)] = size, _a["k-rounded-" + (kendo_vue_common_1.kendoThemeMaps.roundedMap[tagsRounded] || tagsRounded)] = tagsRounded, _a["k-chip-" + fillMode] = fillMode, _a["k-chip-" + fillMode + "-base"] = Boolean(fillMode), _a;
    }
  },
  // @ts-ignore
  setup: !isV3 ? undefined : function () {
    var v3 = !!isV3;
    return {
      v3: v3
    };
  },
  methods: {
    onTagDelete: function onTagDelete(tagData, event) {
      this.$emit('tagdelete', tagData, event);
    }
  },
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
        dataItems = _a.dataItems,
        guid = _a.guid;
    return h("div", {
      "class": this.wrapperClass,
      role: "listbox",
      attrs: this.v3 ? undefined : {
        role: "listbox",
        id: 'tagslist-' + guid
      },
      id: 'tagslist-' + guid
    }, [dataItems.map(function (tagData, index) {
      var _this = this;

      var uniqueId = "tag-" + guid + "-" + tagData.text.toString().replace(/\s+/g, '-');
      var defaultRendering = h("div", {
        "class": __assign(__assign({}, this.chipClass), {
          'k-focus': tagData === this.$props.focused
        }),
        key: uniqueId,
        id: uniqueId,
        attrs: this.v3 ? undefined : {
          id: uniqueId,
          "aria-selected": true,
          role: "option",
          "aria-setsize": dataItems.length
        },
        onMousedown: preventDefault,
        on: this.v3 ? undefined : {
          "mousedown": preventDefault,
          "click": stopPropagation
        },
        onClick: stopPropagation,
        "aria-selected": true,
        role: "option",
        "aria-setsize": dataItems.length
      }, [h("span", {
        "class": "k-chip-content"
      }, [tagData.text]), h("span", {
        "aria-label": "delete",
        attrs: this.v3 ? undefined : {
          "aria-label": "delete"
        },
        "class": "k-select",
        onClick: function onClick(e) {
          return _this.onTagDelete(tagData.data, e);
        },
        on: this.v3 ? undefined : {
          "click": function onClick(e) {
            return _this.onTagDelete(tagData.data, e);
          }
        }
      }, [h("span", {
        "class": "k-icon k-i-close"
      })])]);
      return kendo_vue_common_1.getTemplate.call(this, {
        h: h,
        template: this.$props.tagRender,
        defaultRendering: defaultRendering,
        additionalProps: __assign(__assign({}, this.$props), {
          tagData: tagData,
          index: index
        }),
        additionalListeners: {
          tagdelete: this.onTagDelete
        }
      });
    }, this)]);
  }
};
exports.TagListVue2 = TagListVue2;
/**
 * @hidden
 */

var TagList = TagListVue2;
exports.TagList = TagList;
